### PR TITLE
Append Lua backtrace to C++ exceptions when using regular Lua instead of LuaJIT

### PR DESCRIPTION
**Pull request by celeron55**
~~I am considering this good to go after some testing. Comparisons between Lua and LuaJIT with and without this patch need to be made.~~

The practical tests pass, but there's still at least a race condition issue.

### CODE REVIEW

**Comment by celeron55** [comment]
I found a deadlock while testing this and fixed it. I guess this might be fine now.

**Comment by kwolekr** [comment]
Alright, first off, in the second commit you didn't fix a deadlock, you fixed a race condition introduced by your first commit by locking envlock.  I don't see why you'd use envlock and not scriptlock, however.
Second, I don't really approve of the way this commit 'fixes' the problem.  Sure, it's pragmatic, but it ignores the deeper issue of why plain Lua does not produce backtraces where it logically should.
Sorry celeron... :-1:

**Comment by celeron55** [comment]
"it ignores the deeper issue of why plain Lua does not produce backtraces where it logically should."

Eh, what? I explained this on IRC when I was making this. Regular Lua does not have a wrapper functionality around calls to C functions. That would be unnecessary because exceptions easily pass down through regular Lua unlike LuaJIT's runtime-generated code.

LuaJIT is the one here that isn't pragmatic, not Lua.

**Comment by celeron55** [comment]
I didn't realize there exists a functioning separate scriptlock now though, I'll have to fix that for sure.

**Comment by ShadowNinja** [comment]
In some places you supply an exception handler for `ProcessedLuaError` which is just a copy of the `LuaError` handler.  Sonce PLE inherits from LE a single LE handler should do.

File: src/server.cpp [hunk 1]
```diff
@@ -2846,6 +2848,36 @@ void Server::notifyPlayer(const char *name, const std::wstring &msg)
...
+
+	// LuaJIT provides a way to catch exceptions directly at the C++->Lua
+	// interface and we invoke regular Lua error handling in there, which
+	// appends a backtrace to the error and converts it to a ProcessedLuaError.
+
+	std::ostringstream os(std::ios::binary);
+	os << "Lua: " << error;
+	os << std::endl << m_script->getBacktrace();
+
+	setAsyncFatalError(os.str());
```
Comment (ShadowNinja) [inline src/server.cpp]: A `std:ostringstream` is overkill here.  Just do something like
```C++
std::string err = "Lua: " + error + '\n' + m_script->getBacktrace();
```

**Comment by paramat** [comment]
Needs work and rebase and only 3 days to freeze and does not seem a prioroty for release.

**Comment by rubenwardy** [comment]
no response from author

**Pull request closed without merging**